### ISSUE TITLE
"Hello World" Training

### ISSUE THREAD

**Issue by MaxiBoether** [eth-easl/modyn#51]
In this issue, we track the necessary steps to reach our "Hello World" goal in January. I deliberately make this a separate issue from #34 because #34 comprises additional steps that will be necessary but not for a hello world-functional example.

The goal is to have a simple training job run in the system. This training does not really make lots of sense, but it's enough to be considered a success. The goal is that the following setting should be fully functional.

- We have timestamped MNIST data in storage (can be just increasing timestamps starting at 0, doesn't really matter, just need some form of temporality)
- We train a ResNet18 on that MNIST data to classify the 10 digits. We pre-train on 50% of the data and then replay the other 50%, i.e., the focus lies on the _experiment mode_ of the system, not the production mode
- Our policy is to finetune on all new data, the trigger is every `n` new data points
- The supervisor starts that training according to the pipeline policy given via command line, it fetches the data
- We collect the loss of each data point as metadata

In the best case, we also have an evaluation component that evaluates model performance (in an experiment sense, see #50), but we might want to do this in the next sprint depending on our progress.

Overall, I currently see the following subtasks by component:

**Storage**
- [x] ~~Separate storage and filling of storage. Currently, they are somehow intertwined. We might want separate scripts to push data into storage, but they need to be separated from the storage component itself. Currently, we for example have these DataSources that break with this paradigm. It should be like on the architectural slide: We have some form of storage (in the basic case, just a filesystem), some external application (running in another container!) pushes data into the storage, our storage component notices that and updates its database etc.~~ (storage done in #55, filling of filesystems #80 )
- [x] ~~Refactor the `Storage` component to have a  better structure~~ (partly #11, partly #33). (done in #55) 
- [x] ~~Support dataset ID in storage (#44). The scripts that push data into storage (again, these should be completely separate from the running storage container itself) for a specific dataset.~~ (done in #55)
- [x] ~~Support a "GetNewDataSinceX" function. Given a dataset, it returns all new keys that have arrived since that point in time.~~ (done in #55)

**GPU Node**
- [x] Clearly separate frontend, i.e., our PyTorch wrappers etc, and the backend GPU node, i.e., the gRPC server that receives tasks
- [x] Implement gRPC server waiting for training requests
- [x] Implement generic PyTorch training loop that gets populated with training details from gRPC request. Especially interesting: The preprocessing pipeline (see Supervisor/GPU Node communication).
- [x] Implement MetadataCollector #16 that sends metadata to the MetadataProcessor
- [x] We probably skip validation sets/test sets for the hello world milestone because it's a larger problem to tackle (#50)
- [x] Maxi: Also review the stuff that is already there but might not be touched (e.g., `DynamicDataset`).
- [ ] Remove frontend/backend structure after merging Foteini's initial PR.

**Supervisor**
- [x] ~~Sketch out the supervisor framework~~ (done in #45)
- [x] ~~Write a "Hello World" pipeline policy yaml file that reflects the training job describes above~~ (done in #62)
- [x] ~~Write the pipeline schema (#14)~~ (done in #62)
- [x] ~~Implement a "wait for trigger" functionality by pulling latest information from storage and passing that to trigger~~  (done in #60)
- [x] ~~Write trigger that triggers training on the `n`th new data point~~ (done in #60)

**Supervisor / GPU Node Interaction**
- [x] Define a communication protocol between supervisor and GPU node, i.e., the gRPC requests that get send from supervisor to GPU node
- [x] On trigger, send training request according to communication protocol to GPU node.
- [x] Important subtask of communication protocol: Preprocessing pipeline for images
- [x] Define how the final model gets sent back to the supervisor / how the supervisor pulls the final model
- [x] Make sure that the training is synchronized to the trigger, i.e., we should exactly take the view up until the data point that triggered the training request and do not consider less or more data points that came in
- [x] Transfer training log / live training status from GPU node to supervisor to see progress
- [x] Think about transferring the training log (e.g., validation accuracys)  from gpu node to supervisor. Maybe we don't want/need this yet.

**Selector**
- [x] ~~Clean up/refactor current selector implementation~~ (fixed in #20)
- [x] #64: Handle control flow and synchronization with supervisor. After #46 is merged, the Supervisor and Selector both still function separately. They might have different notions of time and the Selector is not aware of training triggers. This should be synchronized with the Supervisor being the main component. This will also solve the following three TODOs. 
- [x] Support both offline and online algorithms conceptually
- [x] Support pre-training before the actual algorithm implementation
- [x] Support experiment mode (replay of data)

**Metadata**
- [x] ~~Build MetadataDatabase backend component~~ (Fixed in #96)
- [x] Build MetadataProcessor backend component (#13)

**General**
- [x] ~~Get rid of NewQueue in entire repo~~ (Fixed in #53) 
- [x] ~~Get rid of ODM stuff in entire repository~~ (Fixed in #46)
- [x] ~~Update and fix Docker/docker-compose files (#43)~~ (fixed in #67 and #73)
- [x] Implement ResNet18 (will be fixed when #56 is merged)
- [x] #33 Is applicable to all legacy components that we touch here (fixed when #46 and #65 are merged)
- [ ] Add all components to the 90% coverage check action before merging (after #56, #46, and #65 are merged, we can switch the job to include coverage checks for the entire repo. After that is done and fixed - because for example currently the utils folder is not tested - this is solved)
